### PR TITLE
Fix astype for complex dtypes

### LINKED ISSUE #1276: Casting complex to float causes a compilation error
NumPy supports casing complex to float but CuPy does not.

```
>>> numpy.array([1], dtype=numpy.complex64).astype('f')
array([ 1.], dtype=float32)
```

### PR DESCRIPTION
**Pull request by unnonouno**
- complex to bool checks if each element is non-zero
- complex to int and complex to float uses real part

```
>>> numpy.array([0+1j]).astype('?')
array([ True], dtype=bool)
>>> numpy.array([0+1j]).astype('i')
__main__:1: ComplexWarning: Casting complex values to real discards the imaginary part
array([0], dtype=int32)
>>> numpy.array([0+1j]).astype('f')
array([ 0.], dtype=float32)
```

fix #1276

### CODE REVIEW

**Comment by toslunar** [comment]
numpy document (https://docs.scipy.org/doc/numpy/reference/generated/numpy.ndarray.astype.html) says:
> **ComplexWarning** When casting from complex to float or int. To avoid this, one should use a.real.astype(t)."

It's warned only once by default.
```
>>> import numpy
>>> numpy.array([0+1j]).astype('f')
__main__:1: ComplexWarning: Casting complex values to real discards the imaginary part
array([0.], dtype=float32)
>>> numpy.array([0+1j]).astype('i')
array([0], dtype=int32)
```

File: tests/cupy_tests/core_tests/test_ndarray_copy_and_view.py [hunk 1]
```diff
@@ -69,12 +69,17 @@ def test_transposed_fill(self, xp, dtype):
         return b
 
     @testing.for_orders(['C', 'F', 'A', 'K', None])
-    @testing.for_all_dtypes(name='src_dtype', no_complex=True)
+    @testing.for_all_dtypes(name='src_dtype')
```
Comment (toslunar) [inline tests/cupy_tests/core_tests/test_ndarray_copy_and_view.py]: Why don't you use `for_all_dtypes_combination`?

**Comment by unnonouno** [comment]
Yes. It is default behavior of `warnings`.
With this branch, cupy shows a warning once.
```
>>> cupy.array([1], cupy.comple.astype('f')
__main__:1: ComplexWarning: Casting complex values to real discards the imaginary part
array([ 1.], dtype=float32)
>>> cupy.array([1], cupy.complex64).astype('f')
array([ 1.], dtype=float32)
```

**Comment by unnonouno** [comment]
I made a utility function and replace it with all `astype` in the test.

**Comment by okuta** [comment]
@unnonouno Cloud you check test result?

**Comment by unnonouno** [comment]
I found with a recent PR breaks this PR. I fixed it.

**Comment by kmaehashi** [comment]
Jenkins, test this please.

File: cupy/core/core.pyx [hunk 2]
```diff
@@ -356,7 +356,7 @@ cdef class ndarray:
             warnings.warn(
                 'Casting complex values to real discards the imaginary part',
                 numpy.ComplexWarning)
-            elementwise_copy(self.real, newarray)
+            elementwise_copy(ascontiguousarray(self).real, newarray)
```
Comment (toslunar) [inline cupy/core/core.pyx]: I forgot to fix `.real` and `.imag` in #1287

**Comment by toslunar** [comment]
Let me merge the PR since the tests are passing.  I'll fix the issue soon.

**Comment by unnonouno** [comment]
Ah, ok. As you say, it is a problem of `real`.

**Pull request merged**